### ISSUE TITLE
Certain character setups crash clients on load.

### ISSUE THREAD

**Issue by Mewchild** [VOREStation/VOREStation#2186]
#### Brief description of the issue
Still researching the issue, but from in round experience, certain character sprites/set-ups crash players on view.

#### Steps to reproduce
Only evidence thus far is anecdotal, still trying to find a way to reliably replicate
[Character in question](https://i.imgur.com/3k7bIes.png)

#### Code Revision
Server revision: release - 
30d353407955f4849d06c98ef496862de920c346 )
 
#### Anything else you may wish to add:
This mirrors a similar bug we had on Eros station, which was traced back to Driders/Spider Taurs. I believe for us it was something to do with how the lower body was set up (which is mirrored 
by a legless character crashing users). Will continue to research, as stated above.

**Comment by OrbisAnima** [comment]
This is most likely from polaris, given that the Spider taurs you use make use of organ legs, and the bug you are researching is related to removing legs and trying to join. You should report it to them too, so they know there is a bug related to it

**Comment by Mewchild** [comment]
Did some amount of further testing. The issue is definitely linked to the legs being gone. Playing a few rounds with the mechanical arm both there/not there did not alter the results, people still crashed. Not quite sure causes it, or where to begin looking. I've given Polaris a prod in Host Chat, I'll see if they find anything.

**Comment by Mewchild** [comment]
https://i.imgur.com/MKqK5Zr.png
Spoke with Anewbie from Polaris, they've never seen the bug on their end.

**Comment by OrbisAnima** [comment]
Look, the crashes come from making a refference to things that arent there, ie, the legs. And others crash when they are on rendering range, so its likely its a visual issue... so check the markings, its something vore has that polaris doesnt, and even if they were made with base on the damage system, on a normal run they would never print on legs, becouse having no legs mean they are on damage 0. 

TL DR: Check the markings code logic, figure out a way to only interact limbs that exist, that should fix both the Virgo issue and the Eros taurs issue.

**Comment by Cameron653** [comment]
I spent an hour fiddling with this. 
I spawned in with full body markings, one leg, no legs, one leg and a robotic arm (the same as the screenshot), a spider taur half with one leg, spider taur half with no legs.

I did droplimb() on the legs, feet, arm, and sliced the arm and legs off with an esword.
I used an alt to run into view while the limb was being cut off, went into view while the limb was already missing... 

I simply can't reproduce this. I'm just going to chalk this up to BYOND being odd and randomly crashing people, as it's known for. 

I'd also say that this isn't even a new bug, either, just some odd variant of the old VV panel bug. Sometimes if you try to open the VV panel on someone, it'll crash your client. Some others are completely fine opening the VV panel on them, however. The VV panel bug has been out there longer than markings have been in the game, so I'm guessing it's some kind of variant of that bug, it's just moved onto something that is visible on examining.

**Comment by Screemonster** [comment]
If it reoccurs, we need to find out what byond version the victims are running. I've seen shit happen depending on which version of _windows_ I'm running. (the client lag when you get more text in while you have some highlighted, for instance, never seemed to bother my win7 laptop, but it's a killer on 8.1)

**Comment by OrbisAnima** [comment]
alternatively, someone could take the time to test the case of spawning as a legless character and entering in the presence of another, installing and uninstalling the client a few times

**Comment by Cameron653** [comment]
Sweet jesus, it's real. It's true.
https://i.imgur.com/r3c68Gt.png
See: Bottom left hand corner. Someone _just_ coming into view in a wheelchair and my game crashing as a result.

**Comment by OrbisAnima** [comment]
From our conversations with Mewchild, one of the current suspects are the foots and the stump code, for some reason you get stumps generated when you have the legs, at that its fine, but then you get stumps when you get the foot cut, and that might be creating issues somehow, but its all assumptions. 

What should be tested is if this is only caused at joining with no legs or this happens normally if they are cut on play.

**Comment by Mewchild** [comment]
Can confirm that people crash, even if the limbs were amputated in round. I'll test spawning a character with legs but not feet later.
[Legs amputated in round](https://i.imgur.com/xyK9EEb.png) (Feet amputated, and then legs, in order)
[Legs amputated in round](https://i.imgur.com/Mkc8jkl.png)  (Leg+foot at the same time)

**Comment by Mewchild** [comment]
Testing again. Foot removed via [character setup](https://i.imgur.com/KpER2eQ.png)
No crashes, from anyone, over a four hour play session using this setup. It's definitely tied to the legs+feet being gone.

**Comment by SpadesNeil** [comment]
Theory: Feet cannot exist because legs do not exist. A null is being read where one should not exist, resulting in a crash.

**Comment by Mewchild** [comment]
For some people. Sometimes. That's what kills me, is that it's never 100% that it happens to anyone.

**Comment by OrbisAnima** [comment]
I wonder if fixing the stump appearing for legs would fix this...

**Comment by Cameron653** [comment]
Did some more testing.
Right foot missing: Nothing
Right leg missing: Nothing
Both legs missing: Nothing
Both legs missing and a prosthetic arm: Crash https://i.imgur.com/NyabMuD.png
I think it _may_ have to do something with the prosthetic arm.

**Comment by Mewchild** [comment]
I've got some skepticism for that. I'm hopping into round as the character in question, with the only limb modifications being the missing legs (normal arms). I've never seen anybody crash like that near any of my other characters with mechanical limbs, so I'll check on the live server.

**Comment by Mewchild** [comment]
[Jesse Seommer, crashed as soon as I came into view](https://i.imgur.com/LJYXxcq.png)
https://i.imgur.com/rt0c5oe.png

It's not related to the prosthetic, at least not in full.

**Comment by Mewchild** [comment]
As something that was discovered in a round today, this bug can crash users even if it's a ghost that's missing limbs. The example in question was a character who had lost all of their limbs due to a HV scene, who then crashed everybody in the bar and hydroponics multiple times before realizing the issue.

tl;dr limbless ghosts can crash people

**Comment by OrbisAnima** [comment]
For the limbless ghost bug, we can dirty patch it with forcing ghosts to use the old non mobs sprites.

**Comment by OrbisAnima** [comment]
One of the posibilities for this bug is bay map hack protection, which would explain most crashes, since most people crash on line on sight, getting the not feet organ to appear.

**Comment by OrbisAnima** [comment]
Nevermind, found the lines that is borking it all forever.
https://github.com/VOREStation/VOREStation/blob/master/code/modules/mob/living/carbon/human/update_icons.dm#L795
https://github.com/VOREStation/VOREStation/blob/master/code/modules/mob/living/carbon/human/update_icons.dm#L796

basically, this little species.get_bodytype(src) is being called when it doesnt exist, and breaking everything forever, so, one of the fixes that stopped the shenanigans from happening and tested it with Bluehaus, was adding a check for it, namely
`(wear_suit.sprite_sheets **&& species.get_bodytype(src)** && wear_suit.sprite_sheets[species.get_bodytype(src)])`

We did notice that in my local he couldnt see the tajara mob or skrell mob when doing that change, it might be an issue with my local hosting, or might be an issue with the change, but that stopped all the crashes.

**Comment by Mewchild** [comment]
Reopening this. Limbless characters still crash clients, even while ghosted. Observed today, on revision a5844f786ef2a09948e6718a7d418b05d95f9324

**Comment by SpadesNeil** [comment]
Luckily the crash only happens once, then you're usually good. Can anyone get their BYOND client crash logs?

**Comment by OrbisAnima** [comment]
....guys, you deleted my fix in the last update.

**Comment by OrbisAnima** [comment]
Tested the issue with Natje, wharever is going on now, doesnt need line of sight like before, instead, it can happen in the normal sight radious as if the walls werent there.

A note when testing the issue, you need to use a new character every time you recreate the issue for the same person to crash.

**Comment by Mewchild** [comment]
Bit an odd question here, but has [this](https://i.imgur.com/jCaji7g.png) been addressed/re-added?

**Comment by Cameron653** [comment]
As far as I can see, his fix is still there?

https://github.com/VOREStation/VOREStation/blob/master/code/modules/mob/living/carbon/human/update_icons.dm#L867

**Comment by OrbisAnima** [comment]
I deleted that comment becouse there was a mixup with the releases in that moment, no idea why you kept that capture...

**Comment by OrbisAnima** [comment]
In any case, the problem that I detected that time was a possibility of null being called upon view, the current bug seems to be different, with the crash not being upon sight, but upon range of detection of the character. Its quite puzzling, becouse it seems to be a different issue, since the trigger is different now.

I also noted that this time, unless the character is changed, you dont get a second crash.

**Comment by Mewchild** [comment]
Not that there was expectation otherwise, but this still exists as of e60fa65165516de6fde793c632f111bf7a9a1836

**Comment by SpadesNeil** [comment]
https://github.com/VOREStation/VOREStation/issues/2660 Possibly related.

**Comment by Mewchild** [comment]
Apparently still an issue as f current release. Will test later and add more details. Is apparently a very slippery BYOND bug

**Comment by Arokha** [comment]
Being a byond bug and all, it's not possible for us to fix it, only figure out what causes it and then just avoid doing that anymore, or try to do it another way.

**Comment by Verkister** [comment]
I've been code diving again to see what makes legs so special and found a somewhat flimsy looking part in the icon cache generation code responsible of layering left/right legs for side views. Made that code part a little less flimsy and it's currently open as Polaris PR. Hopefully we'll see if that had any effect.

**Comment by Arokha** [comment]
http://www.byond.com/forum/?post=2348506

Contribute to this thread (or the many other previous threads) about this issue if you want to help fix it. Keep in mind that it's the byond forums, they have no idea about 'limbs' or anything game-specific.
